### PR TITLE
feat(group): refactor searchGroups to make use of SearchQueryBuilder

### PR DESCRIPTION
**Pull request by jgravois**
* make `searchGroups` and `searchItems` signatures consistent
* add support for SearchQueryBuilder to `searchGroups`
* refactor out a shared internal method to keep things DRY
* deprecate `IGroupSearchRequest` and `IGroupSearchResult` interfaces
* make `ISearchResult` and `ISearchRequestOptions` a little more generic

> disclaimer: skipping `export * "./util/generic-search";` is _not_ enough to ensure the internal utility method is excluded from the API reference.

AFFECTS PACKAGES:
@esri/arcgis-rest-portal

BREAKING CHANGE:
searchGroups signature change

ISSUES CLOSED: #104

### CODE REVIEW

File: packages/arcgis-rest-portal/src/groups/search.ts [hunk 1]
```diff
@@ -1,65 +1,24 @@
...
  * //
- * searchGroups({
- *   q:'water'
- * })
- *   .then(response) // response.results.total => 355
+ * searchGroups('water')
+ *   .then(response) // response.total => 355
  * ```
- * Search a portal for groups. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/group-search.htm) for more information.
+ * Search a portal for items. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/search.htm) for more information.
```
Comment (tomwayson) [inline packages/arcgis-rest-portal/src/groups/search.ts]: shouldn't this be "Search portal for _groups_"?

File: packages/arcgis-rest-portal/src/util/generic-search.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  appendCustomParams
+} from "@esri/arcgis-rest-request";
+
+import { SearchQueryBuilder } from "./SearchQueryBuilder";
+import { getPortalUrl } from "../util/get-portal-url";
+import { ISearchRequestOptions, ISearchResult } from "../util/search";
+
+export function genericSearch(
+  search: string | ISearchRequestOptions | SearchQueryBuilder,
+  searchType: "item" | "group"
```
Comment (patrickarlt) [inline packages/arcgis-rest-portal/src/util/generic-search.ts]: @jgravois this LGTM but I'm going to pull this down and see if I can improve this typing at all.

**Comment by patrickarlt** [comment]
@jgravois I've updated the types for `genericSearch` and `ISearchResponse` to more clearly indicate that they will ONLY return an array of `IGroup` or `IItem`, I had to `npm update` to get typedoc to accept the latest changes so that is why there is a bunch of random package updates.

**Comment by jgravois** [comment]
> I've updated the types for `genericSearch` and `ISearchResponse` to more clearly indicate that they will ONLY return an array of `IGroup` or `IItem`

very cool!

**Pull request merged**